packages/messaging/src/helpers/migrate-old-database.ts

```diff
@@ -0,0 +1,193 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { openDb, deleteDb } from 'idb';
+import { TokenDetails } from '../interfaces/token-details';
+import { arrayToBase64 } from './array-to-base64';
+
+// https://github.com/firebase/firebase-js-sdk/blob/7857c212f944a2a9eb421fd4cb7370181bc034b5/packages/messaging/src/interfaces/token-details.ts
+export interface V2TokenDetails {
+  fcmToken: string;
+  swScope: string;
+  vapidKey: string | Uint8Array;
+  subscription: PushSubscription;
+  fcmSenderId: string;
+  fcmPushSet: string;
+  createTime?: number;
+  endpoint?: string;
+  auth?: string;
+  p256dh?: string;
+}
+
+// https://github.com/firebase/firebase-js-sdk/blob/6b5b15ce4ea3df5df5df8a8b33a4e41e249c7715/packages/messaging/src/interfaces/token-details.ts
+export interface V3TokenDetails {
+  fcmToken: string;
+  swScope: string;
+  vapidKey: Uint8Array;
+  fcmSenderId: string;
+  fcmPushSet: string;
+  endpoint: string;
+  auth: ArrayBuffer;
+  p256dh: ArrayBuffer;
+  createTime: number;
+}
+
+// https://github.com/firebase/firebase-js-sdk/blob/9567dba664732f681fa7fe60f5b7032bb1daf4c9/packages/messaging/src/interfaces/token-details.ts
+export interface V4TokenDetails {
+  fcmToken: string;
+  swScope: string;
+  vapidKey: Uint8Array;
+  fcmSenderId: string;
+  endpoint: string;
+  auth: ArrayBufferLike;
+  p256dh: ArrayBufferLike;
+  createTime: number;
+}
+
+const OLD_DB_NAME = 'fcm_token_details_db';
+/**
+ * The last DB version of 'fcm_token_details_db' was 4. This is one higher,
+ * so that the upgrade callback is called for all versions of the old DB.
+ */
+const OLD_DB_VERSION = 5;
+const OLD_OBJECT_STORE_NAME = 'fcm_token_object_Store';
```

Comment:
Why capital "S"?

Reply:
No idea. It has always been like that and it always bothered me.

https://github.com/firebase/firebase-js-sdk/blob/73a586c92afe3f39a844b2be86086fddb6877bb7/packages/messaging/src/models/token-details-model.ts#L22